### ISSUE TITLE
船舶航行轨迹预测代码

### ISSUE THREAD

**Issue by 676524972** [676524972/hello-world#1]
import numpy as np
import csv
from sklearn.preprocessing import MinMaxScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
from tensorflow.keras.optimizers import Adam

# 数据加载函数
def load_data(filename):
    mmsi = []
    latitudes = []
    longitudes = []
    sog = []
    cog = []

    with open(filename, 'r') as file:
        reader = csv.reader(file)
        next(reader)  # Skip header
        for row in reader:
            mmsi.append(row[0])
            latitudes.append(float(row[1]))
            longitudes.append(float(row[2]))
            sog.append(float(row[3]))
            cog.append(float(row[4]))

    return np.array(mmsi), np.array(latitudes), np.array(longitudes), np.array(sog), np.array(cog)

# 数据预处理
def preprocess_data(lats, lons, sog, cog, window_size=10):
    scaler = MinMaxScaler()

    # 归一化经纬度，速度和角度
    features = np.column_stack((lats, lons, sog, cog))
    features_scaled = scaler.fit_transform(features)

    # 构建时间序列数据
    X, y = [], []
    for i in range(len(features_scaled) - window_size):
        X.append(features_scaled[i:i + window_size])
        y.append(features_scaled[i + window_size])

    return np.array(X), np.array(y), scaler

# LSTM模型构建
def build_model(input_shape):
    model = Sequential()
    model.add(LSTM(64, return_sequences=False, input_shape=input_shape))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(4))  # 4个输出：预测经纬度，速度，航向
    model.compile(optimizer=Adam(), loss='mse')
    return model

# 结果反归一化
def inverse_transform(scaler, pred):
    pred_rescaled = scaler.inverse_transform(pred)
    return pred_rescaled

# 保存预测结果为CSV文件
def save_submission(predictions, mmsi, filename='submission.csv'):
    with open(filename, 'w', newline='') as csvfile:
        fieldnames = ['mmsi', 'lat', 'lon']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()

        for i, m in enumerate(mmsi):
            lat, lon = predictions[i]
            writer.writerow({'mmsi': m, 'lat': lat, 'lon': lon})

# 主函数
def main():
    # 加载训练数据
    mmsi_train, latitudes_train, longitudes_train, sog_train, cog_train = load_data('train.csv')
    X_train, y_train, scaler = preprocess_data(latitudes_train, longitudes_train, sog_train, cog_train, window_size=10)

    # 构建并训练模型
    model = build_model(X_train.shape[1:])
    model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

    # 加载测试数据
    mmsi_test, latitudes_test, longitudes_test, sog_test, cog_test = load_data('test.csv')
    X_test, _, _ = preprocess_data(latitudes_test, longitudes_test, sog_test, cog_test, window_size=10)

    # 使用模型预测
    predictions = model.predict(X_test)

    # 反归一化结果
    pred_rescaled = inverse_transform(scaler, predictions)

    # 只输出预测的经纬度
    pred_lats_lons = pred_rescaled[:, :2]

    # 保存结果
    save_submission(pred_lats_lons, mmsi_test, filename='submission.csv')

if __name__ == '__main__':
    main()
